### PR TITLE
src/ops/validate.rs: Common-up some type row calculations

### PR DESCRIPTION
**Pull request by acl-cqc**
Various of these exist already as methods on the classes in `controlflow.rs`...

### CODE REVIEW

File: src/ops/validate.rs [hunk 1]
```diff
@@ -162,11 +162,10 @@ impl ValidateOp for super::Conditional {
...
             let OpType::Case(case_op) = optype else {panic!("Child check should have already checked valid ops.")};
             let sig = &case_op.signature;
-            let predicate_value = &self.predicate_inputs[i];
-            if sig.input[0..predicate_value.len()] != predicate_value[..]
-                || sig.input[predicate_value.len()..] != self.other_inputs[..]
-                || sig.output != self.outputs
-            {
+            if let Some(expected_inputs) = self.case_input_row(i) {
+                if sig.input == expected_inputs && sig.output == self.outputs {
+                    continue;
```
Comment (acl-cqc) [inline src/ops/validate.rs:167]: I had to do it this way (i.e. inverting the predicate) to use `let Some` as I couldn't find a way to compare an `Option<&TypeRow>` with an `Option<TypeRow>`. Seems like I needed the opposite of `as_ref`. Maybe you can see a way?

Comment (acl-cqc) [reply src/ops/validate.rs:167]: (that is, I wanted to write something like `if Some(sig.input) != &self.case_input_row(i)` but with some `&`s or similar on the LHS but couldn't work out what)

File: src/ops/validate.rs [hunk 2]
```diff
@@ -162,11 +162,10 @@ impl ValidateOp for super::Conditional {
         for (i, (child, optype)) in children.into_iter().enumerate() {
             let OpType::Case(case_op) = optype else {panic!("Child check should have already checked valid ops.")};
             let sig = &case_op.signature;
-            let predicate_value = &self.predicate_inputs[i];
-            if sig.input[0..predicate_value.len()] != predicate_value[..]
-                || sig.input[predicate_value.len()..] != self.other_inputs[..]
-                || sig.output != self.outputs
-            {
+            if let Some(expected_inputs) = self.case_input_row(i) {
```
Comment (aborgna-q) [inline src/ops/validate.rs:165]: Should this be an `unwrap` instead? We know this is a Case op.

**Review by aborgna-q — APPROVED** [review]
👌

File: src/ops/validate.rs (hunk 2 above)
Comment (acl-cqc) [reply src/ops/validate.rs:165]: Hmmm, actually that would work (and solve the problem below too), as the length check above should rule out None

Comment (aborgna-q) [reply src/ops/validate.rs]: Alternatively, I think you could have written
```rust
Some(&sig.input) != self.case_input_row(i).as_ref()
```

**Comment by acl-cqc** [comment]
@cqc-alec I think this PR demonstrates the issue with the coverage check. The check fails, because I have shortened/compacted the code that is tested, but not done anything to the untested code...

**Comment by cqc-alec** [comment]
> @cqc-alec I think this PR demonstrates the issue with the coverage check. The check fails, because I have shortened/compacted the code that is tested, but not done anything to the untested code...

That is an issue yes. I guess we just ignore it in that case.

File: src/ops/validate.rs (hunk 2 above)
Comment (acl-cqc) [reply src/ops/validate.rs]: (Actually the RHS needs to be `(&self.case_input_row(i)).as_ref()` but the key is to put the `as_ref` on the other side, which should have been obvious when I said I needed its inverse...d'oh)

**Pull request merged**